### DISCUSSION TITLE
Create binding for templated methods that don't take any parameters

### DISCUSSION THREAD

**Discussion by JackAshwell11** [pybind/pybind11#4936] [Q&A]
Hi, I'm not sure if this is possible, but here is my situation. I want to create a binding for these methods:
```cpp
  template <typename T>
  void add_system() {
    // Check if the system is already registered
    const std::type_info &system_type{typeid(T)};
    if (systems_.contains(system_type)) {
      throw RegistryError("system", system_type.name(), "is already registered with the registry");
    }

    // Add the system to the registry
    systems_[system_type] = std::make_shared<T>(this);
  }

  template <typename T>
  auto find_system() const -> std::shared_ptr<T> {
    // Check if the system is registered
    auto system_result{systems_.find(typeid(T))};
    if (system_result == systems_.end()) {
      throw RegistryError("system", typeid(T).name());
    }

    // Return the system casting it to T
    return std::static_pointer_cast<T>(system_result->second);
  }
```
Since these methods don't take any parameters, the normal way of doing:
```cpp
m.def("add_system", &Registry::add_system<DamageSystem>)
```
Doesn't seem to work in Python when I try to do:
```py
reg = Registry()
reg.add_system[DamageSystem]()
```
So is there a way to create a binding for this or does the method need a parameter?

P.S. Here is my binding code for `DamageSystem` if you needed it:
```cpp
  pybind11::class_<DamageSystem, SystemBase>(systems, "DamageSystem")
      .def(pybind11::init<Registry *>(), pybind11::arg("registry"))
      .def("deal_damage", &DamageSystem::deal_damage);
```

**Comment by evguran** [comment]
Hi. 
1. Compiler converts template functions to its instance on compiling, so pythons `add_system` is equal to cpp `add_system<DamageSystem>`. What kind of error pythons throws? Why do you use [] operator here `reg.add_system[DamageSystem]()`?
2. Judjing the code, add_system is a class method, but you define it as a modules function, so it should be static? Then you call it as a method of reg class.. 
So, please, provide more details
